File: proton-c/bindings/ruby/lib/event/event.rb
#--
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#++

module Qpid::Proton

  module Event

    # @private
    def self.event_type(const_name, method_name = nil) # :nodoc:
      unless Cproton.const_defined?(const_name)
        raise RuntimeError.new("no such constant: #{const_name}")
      end

      const_value = Cproton.const_get(const_name)
      method_name = "on_#{const_name.to_s[3..-1]}".downcase if method_name.nil?

      EventType.new(const_value, method_name)
    end

    # Defined as a programming convenience. No even of this type will ever
    # be generated.
    NONE = event_type(:PN_EVENT_NONE)

    # A reactor has been started.
    REACTOR_INIT = event_type(:PN_REACTOR_INIT)
    # A reactor has no more events to process.
    REACTOR_QUIESCED = event_type(:PN_REACTOR_QUIESCED)
    # A reactor has been stopred.
    REACTOR_FINAL = event_type(:PN_REACTOR_FINAL)

    # A timer event has occurred.
    TIMER_TASK = event_type(:PN_TIMER_TASK)

    # A connection has been created. This is the first even that will ever
    # be issued for a connection.
    CONNECTION_INIT = event_type(:PN_CONNECTION_INIT)
    # A conneciton has been bound toa  transport.
    CONNECTION_BOUND = event_type(:PN_CONNECTION_BOUND)
    # A connection has been unbound from its transport.
    CONNECTION_UNBOUND = event_type(:PN_CONNECTION_UNBOUND)
    # A local connection endpoint has been opened.
    CONNECTION_LOCAL_OPEN = event_type(:PN_CONNECTION_LOCAL_OPEN)
    # A local connection endpoint has been closed.
    CONNECTION_LOCAL_CLOSE = event_type(:PN_CONNECTION_LOCAL_CLOSE)
    # A remote endpoint has opened its connection.
    CONNECTION_REMOTE_OPEN = event_type(:PN_CONNECTION_REMOTE_OPEN)
    # A remote endpoint has closed its connection.
    CONNECTION_REMOTE_CLOSE = event_type(:PN_CONNECTION_REMOTE_CLOSE)
    # A connection has been freed and any outstanding processing has been
    # completed. This is the final event htat will ever be issued for a
    # connection
    CONNECTION_FINAL = event_type(:PN_CONNECTION_FINAL)

    # A session has been created. This is the first event that will ever be
    # issues for a session.
    SESSION_INIT = event_type(:PN_SESSION_INIT)
    # A local session endpoint has been opened.
    SESSION_LOCAL_OPEN = event_type(:PN_SESSION_LOCAL_OPEN)
    # A local session endpoint has been closed.
    SESSION_LOCAL_CLOSE = event_type(:PN_SESSION_LOCAL_CLOSE)
    # A remote endpoint has opened its session.
    SESSION_REMOTE_OPEN = event_type(:PN_SESSION_REMOTE_OPEN)
    # A remote endpoint has closed its session.
    SESSION_REMOTE_CLOSE = event_type(:PN_SESSION_REMOTE_CLOSE)
    # A session has been freed and any outstanding processing has been
    # completed. This is the final event that will ever be issued for a
    # session
    SESSION_FINAL = event_type(:PN_SESSION_FINAL)

    # A link has been created. This is the first event that will ever be
    # issued for a link.
    LINK_INIT = event_type(:PN_LINK_INIT)
    # A local link endpoint has been opened.
    LINK_LOCAL_OPEN = event_type(:PN_LINK_LOCAL_OPEN)
    # A local link endpoint has been closed.
    LINK_LOCAL_CLOSE = event_type(:PN_LINK_LOCAL_CLOSE)
    # A local link endpoint has been detached.
    LINK_LOCAL_DETACH = event_type(:PN_LINK_LOCAL_DETACH)
    # A remote endpoint has opened its link.
    LINK_REMOTE_OPEN = event_type(:PN_LINK_REMOTE_OPEN)
    # A remote endpoint has closed its link.
    LINK_REMOTE_CLOSE = event_type(:PN_LINK_REMOTE_CLOSE)
    # A remote endpoint has detached its link.
    LINK_REMOTE_DETACH = event_type(:PN_LINK_REMOTE_DETACH)
    # The flow control state for a link has changed.
    LINK_FLOW = event_type(:PN_LINK_FLOW)
    # A link has been freed and any outstanding processing has been completed.
    # This is the final event htat will ever be issued for a link.
    LINK_FINAL = event_type(:PN_LINK_FINAL)

    # A delivery has been created or updated.
    DELIVERY = event_type(:PN_DELIVERY)

    # A transport has new data to read and/or write.
    TRANSPORT = event_type(:PN_TRANSPORT)
    # Indicates that a transport error has occurred.
    # @see Transport#condition To access the details of the error.
    TRANSPORT_ERROR = event_type(:PN_TRANSPORT_ERROR)
    # Indicates that the head of a transport has been closed. This means the
    # transport will never produce more bytes for output to the network.
    TRANSPORT_HEAD_CLOSED = event_type(:PN_TRANSPORT_HEAD_CLOSED)
    # Indicates that the trail of a transport has been closed. This means the
    # transport will never be able to process more bytes from the network.
    TRANSPORT_TAIL_CLOSED = event_type(:PN_TRANSPORT_TAIL_CLOSED)
    # Indicates that both the head and tail of a transport are closed.
    TRANSPORT_CLOSED = event_type(:PN_TRANSPORT_CLOSED)

    SELECTABLE_INIT = event_type(:PN_SELECTABLE_INIT)
    SELECTABLE_UPDATED = event_type(:PN_SELECTABLE_UPDATED)
    SELECTABLE_READABLE = event_type(:PN_SELECTABLE_READABLE)
    SELECTABLE_WRITABLE = event_type(:PN_SELECTABLE_WRITABLE)
    SELECTABLE_EXPIRED = event_type(:PN_SELECTABLE_EXPIRED)
    SELECTABLE_ERROR = event_type(:PN_SELECTABLE_ERROR)
    SELECTABLE_FINAL = event_type(:PN_SELECTABLE_FINAL)

    # An Event provides notification of a state change within the protocol
    # engine.
    #
    # Every event has a type that identifies what sort of state change has
    # occurred, along with a pointer to the object whose state has changed,
    # and also any associated objects.
    #
    # For more details on working with Event, please refer to Collector.
    #
    # @see Qpid::Proton::Event The list of predefined events.
    #
    class Event < EventBase

      # @private
      include Qpid::Proton::Util::ClassWrapper
      # @private
      include Qpid::Proton::Util::Wrapper

      # Creates a Ruby object for the given pn_event_t.
      #
      # @private
      def self.wrap(impl, number = nil)
        return nil if impl.nil?

        result = self.fetch_instance(impl, :pn_event_attachments)
        return result unless result.nil?
        number = Cproton.pn_event_type(impl) if number.nil?
        event = Event.new(impl, number)
        return event.context if event.context.is_a? EventBase
        return event
      end

      # @private
      def initialize(impl, number)
        @impl = impl
        class_name = Cproton.pn_class_name(Cproton.pn_event_class(impl))
        context = class_wrapper(class_name, Cproton.pn_event_context(impl))
        event_type = EventType.by_type(Cproton.pn_event_type(impl))
        super(class_name, context, event_type)
        @type = EventType.by_type(number)
        self.class.store_instance(self, :pn_event_attachments)
      end

      # Notifies the handler(s) of this event.
      #
      # If a handler responds to the event's method then that method is invoked
      # and passed the event. Otherwise, if the handler defines the
      # +on_unhandled+ method, then that will be invoked instead.
      #
      # If the handler defines a +handlers+ method then that will be invoked and
      # passed the event afterward.
      #
      # @example
      #
      #   class FallbackEventHandler
      #
      #     # since it now defines a handlers method, any event will iterate
      #     # through them and invoke the +dispatch+ method on each
      #     attr_accessor handlers
      #
      #     def initialize
      #       @handlers = []
      #     end
      #
      #     # invoked for any event not otherwise handled
      #     def on_unhandled(event)
      #       puts "Unable to invoke #{event.type.method} on #{event.context}."
      #     end
      #
      #   end
      #
      # @param handler [Object] An object which implements either the event's
      #    handler method or else responds to :handlers with an array of other
      #    handlers.
      #
      def dispatch(handler, type = nil)
        type = @type if type.nil?
        if handler.is_a?(Qpid::Proton::Handler::WrappedHandler)
          Cproton.pn_handler_dispatch(handler.impl, @impl, type.number)
        else
          result = Qpid::Proton::Event.dispatch(handler, type.method, self)
          if (result != "DELEGATED") && handler.respond_to?(:handlers)
            handler.handlers.each do |hndlr|
              self.dispatch(hndlr)
            end
          end
        end
      end

      # Returns the reactor for this event.
      #
      # @return [Reactor, nil] The reactor.
      #
      def reactor
        impl = Cproton.pn_event_reactor(@impl)
        Qpid::Proton::Util::ClassWrapper::WRAPPERS["pn_reactor"].call(impl)
      end

      def container
        impl = Cproton.pn_event_reactor(@impl)
        Qpid::Proton::Util::ClassWrapper::WRAPPERS["pn_reactor"].call(impl)
      end

      # Returns the transport for this event.
      #
      # @return [Transport, nil] The transport.
      #
      def transport
        Qpid::Proton::Transport.wrap(Cproton.pn_event_transport(@impl))
      end

      # Returns the Connection for this event.
      #
      # @return [Connection, nil] The connection.
      #
      def connection
        Qpid::Proton::Connection.wrap(Cproton.pn_event_connection(@impl))
      end

      # Returns the Session for this event.
      #
      # @return [Session, nil] The session
      #
      def session
        Qpid::Proton::Session.wrap(Cproton.pn_event_session(@impl))
      end

      # Returns the Link for this event.
      #
      # @return [Link, nil] The link.
      #
      def link
        Qpid::Proton::Link.wrap(Cproton.pn_event_link(@impl))
      end

      # Returns the Sender, or nil if there is no Link, associated  with this
      # event if that link is a sender.
      #
      # @return [Sender, nil] The sender.
      #
      def sender
        return self.link if !self.link.nil? && self.link.sender?
      end

      # Returns the Receiver, or nil if there is no Link, associated with this
      # event if that link is a receiver.
      #
      # @return [Receiver, nil] The receiver.
      #
      def receiver
        return self.link if !self.link.nil? && self.link.receiver?
      end

      # Returns the Delivery associated with this event.
      #
      # @return [Delivery, nil] The delivery.
      #
      def delivery
        Qpid::Proton::Delivery.wrap(Cproton.pn_event_delivery(@impl))
      end

      # Sets the message.
      #
      # @param message [Qpid::Proton::Message] The message
      #
      def message=(message)
        @message = message
      end

      # Returns the message.
      #
      # @return [Qpid::Proton::Message] The message.
      #
      def message
        @message
      end

      # @private
      def to_s
        "#{self.type}(#{self.context})"
      end

    end

  end

end

File: proton-c/src/core/event.c
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
#include <stdio.h>
#include <proton/object.h>
#include <proton/event.h>
#include <proton/reactor.h>
#include <assert.h>

struct pn_collector_t {
  pn_list_t *pool;
  pn_event_t *head;
  pn_event_t *tail;
  bool freed;
  bool head_returned;         /* Head has been returned by pn_collector_next() */
};

struct pn_event_t {
  pn_list_t *pool;
  const pn_class_t *clazz;
  void *context;    // depends on clazz
  pn_record_t *attachments;
  pn_event_t *next;
  pn_event_type_t type;
};

static void pn_collector_initialize(pn_collector_t *collector)
{
  collector->pool = pn_list(PN_OBJECT, 0);
  collector->head = NULL;
  collector->tail = NULL;
  collector->freed = false;
}

static void pn_collector_drain(pn_collector_t *collector)
{
  assert(collector);
  while (pn_collector_next(collector))
    ;
  assert(!collector->head);
  assert(!collector->tail);
}

static void pn_collector_shrink(pn_collector_t *collector)
{
  assert(collector);
  pn_list_clear(collector->pool);
}

static void pn_collector_finalize(pn_collector_t *collector)
{
  pn_collector_drain(collector);
  pn_decref(collector->pool);
}

static int pn_collector_inspect(pn_collector_t *collector, pn_string_t *dst)
{
  assert(collector);
  int err = pn_string_addf(dst, "EVENTS[");
  if (err) return err;
  pn_event_t *event = collector->head;
  bool first = true;
  while (event) {
    if (first) {
      first = false;
    } else {
      err = pn_string_addf(dst, ", ");
      if (err) return err;
    }
    err = pn_inspect(event, dst);
    if (err) return err;
    event = event->next;
  }
  return pn_string_addf(dst, "]");
}

#define pn_collector_hashcode NULL
#define pn_collector_compare NULL

PN_CLASSDEF(pn_collector)

pn_collector_t *pn_collector(void)
{
  return pn_collector_new();
}

void pn_collector_free(pn_collector_t *collector)
{
  assert(collector);
  pn_collector_release(collector);
  pn_decref(collector);
}

void pn_collector_release(pn_collector_t *collector)
{
  assert(collector);
  if (!collector->freed) {
    collector->freed = true;
    pn_collector_drain(collector);
    pn_collector_shrink(collector);
  }
}

pn_event_t *pn_event(void);

pn_event_t *pn_collector_put(pn_collector_t *collector,
                             const pn_class_t *clazz, void *context,
                             pn_event_type_t type)
{
  if (!collector) {
    return NULL;
  }

  assert(context);

  if (collector->freed) {
    return NULL;
  }

  pn_event_t *tail = collector->tail;
  if (tail && tail->type == type && tail->context == context) {
    return NULL;
  }

  clazz = clazz->reify(context);

  pn_event_t *event = (pn_event_t *) pn_list_pop(collector->pool);

  if (!event) {
    event = pn_event();
  }

  event->pool = collector->pool;
  pn_incref(event->pool);

  if (tail) {
    tail->next = event;
    collector->tail = event;
  } else {
    collector->tail = event;
    collector->head = event;
  }

  event->clazz = clazz;
  event->context = context;
  event->type = type;
  pn_class_incref(clazz, event->context);

  return event;
}

pn_event_t *pn_collector_peek(pn_collector_t *collector)
{
  return collector->head;
}

bool pn_collector_pop(pn_collector_t *collector)
{
  collector->head_returned = false;
  pn_event_t *event = collector->head;
  if (event) {
    collector->head = event->next;
  } else {
    return false;
  }

  if (!collector->head) {
    collector->tail = NULL;
  }

  pn_decref(event);
  return true;
}

pn_event_t *pn_collector_next(pn_collector_t *collector)
{
  if (collector->head_returned) {
    pn_collector_pop(collector);
  }
  collector->head_returned = collector->head;
  return collector->head;
}

pn_event_t *pn_collector_prev(pn_collector_t *collector) {
  return collector->head_returned ? collector->head : NULL;
}

bool pn_collector_more(pn_collector_t *collector)
{
  assert(collector);
  return collector->head && collector->head->next;
}

static void pn_event_initialize(pn_event_t *event)
{
  event->pool = NULL;
  event->type = PN_EVENT_NONE;
  event->clazz = NULL;
  event->context = NULL;
  event->next = NULL;
  event->attachments = pn_record();
}

static void pn_event_finalize(pn_event_t *event) {
  // decref before adding to the free list
  if (event->clazz && event->context) {
    pn_class_decref(event->clazz, event->context);
  }

  pn_list_t *pool = event->pool;

  if (pool && pn_refcount(pool) > 1) {
    event->pool = NULL;
    event->type = PN_EVENT_NONE;
    event->clazz = NULL;
    event->context = NULL;
    event->next = NULL;
    pn_record_clear(event->attachments);
    pn_list_add(pool, event);
  } else {
    pn_decref(event->attachments);
  }

  pn_decref(pool);
}

static int pn_event_inspect(pn_event_t *event, pn_string_t *dst)
{
  assert(event);
  assert(dst);
  const char *name = pn_event_type_name(event->type);
  int err;
  if (name) {
    err = pn_string_addf(dst, "(%s", pn_event_type_name(event->type));
  } else {
    err = pn_string_addf(dst, "(<%u>", (unsigned int) event->type);
  }
  if (err) return err;
  if (event->context) {
    err = pn_string_addf(dst, ", ");
    if (err) return err;
    err = pn_class_inspect(event->clazz, event->context, dst);
    if (err) return err;
  }

  return pn_string_addf(dst, ")");
}

#define pn_event_hashcode NULL
#define pn_event_compare NULL

PN_CLASSDEF(pn_event)

pn_event_t *pn_event(void)
{
  return pn_event_new();
}

pn_event_type_t pn_event_type(pn_event_t *event)
{
  return event->type;
}

const pn_class_t *pn_event_class(pn_event_t *event)
{
  assert(event);
  return event->clazz;
}

void *pn_event_context(pn_event_t *event)
{
  assert(event);
  return event->context;
}

pn_record_t *pn_event_attachments(pn_event_t *event)
{
  assert(event);
  return event->attachments;
}

const char *pn_event_type_name(pn_event_type_t type)
{
  switch (type) {
  case PN_EVENT_NONE:
    return "PN_EVENT_NONE";
  case PN_REACTOR_INIT:
    return "PN_REACTOR_INIT";
  case PN_REACTOR_QUIESCED:
    return "PN_REACTOR_QUIESCED";
  case PN_REACTOR_FINAL:
    return "PN_REACTOR_FINAL";
  case PN_TIMER_TASK:
    return "PN_TIMER_TASK";
  case PN_CONNECTION_INIT:
    return "PN_CONNECTION_INIT";
  case PN_CONNECTION_BOUND:
    return "PN_CONNECTION_BOUND";
  case PN_CONNECTION_UNBOUND:
    return "PN_CONNECTION_UNBOUND";
  case PN_CONNECTION_REMOTE_OPEN:
    return "PN_CONNECTION_REMOTE_OPEN";
  case PN_CONNECTION_LOCAL_OPEN:
    return "PN_CONNECTION_LOCAL_OPEN";
  case PN_CONNECTION_REMOTE_CLOSE:
    return "PN_CONNECTION_REMOTE_CLOSE";
  case PN_CONNECTION_LOCAL_CLOSE:
    return "PN_CONNECTION_LOCAL_CLOSE";
  case PN_CONNECTION_FINAL:
    return "PN_CONNECTION_FINAL";
  case PN_SESSION_INIT:
    return "PN_SESSION_INIT";
  case PN_SESSION_REMOTE_OPEN:
    return "PN_SESSION_REMOTE_OPEN";
  case PN_SESSION_LOCAL_OPEN:
    return "PN_SESSION_LOCAL_OPEN";
  case PN_SESSION_REMOTE_CLOSE:
    return "PN_SESSION_REMOTE_CLOSE";
  case PN_SESSION_LOCAL_CLOSE:
    return "PN_SESSION_LOCAL_CLOSE";
  case PN_SESSION_FINAL:
    return "PN_SESSION_FINAL";
  case PN_LINK_INIT:
    return "PN_LINK_INIT";
  case PN_LINK_REMOTE_OPEN:
    return "PN_LINK_REMOTE_OPEN";
  case PN_LINK_LOCAL_OPEN:
    return "PN_LINK_LOCAL_OPEN";
  case PN_LINK_REMOTE_CLOSE:
    return "PN_LINK_REMOTE_CLOSE";
  case PN_LINK_LOCAL_DETACH:
    return "PN_LINK_LOCAL_DETACH";
  case PN_LINK_REMOTE_DETACH:
    return "PN_LINK_REMOTE_DETACH";
  case PN_LINK_LOCAL_CLOSE:
    return "PN_LINK_LOCAL_CLOSE";
  case PN_LINK_FLOW:
    return "PN_LINK_FLOW";
  case PN_LINK_FINAL:
    return "PN_LINK_FINAL";
  case PN_DELIVERY:
    return "PN_DELIVERY";
  case PN_TRANSPORT:
    return "PN_TRANSPORT";
  case PN_TRANSPORT_AUTHENTICATED:
    return "PN_TRANSPORT_AUTHENTICATED";
  case PN_TRANSPORT_ERROR:
    return "PN_TRANSPORT_ERROR";
  case PN_TRANSPORT_HEAD_CLOSED:
    return "PN_TRANSPORT_HEAD_CLOSED";
  case PN_TRANSPORT_TAIL_CLOSED:
    return "PN_TRANSPORT_TAIL_CLOSED";
  case PN_TRANSPORT_CLOSED:
    return "PN_TRANSPORT_CLOSED";
  case PN_SELECTABLE_INIT:
    return "PN_SELECTABLE_INIT";
  case PN_SELECTABLE_UPDATED:
    return "PN_SELECTABLE_UPDATED";
  case PN_SELECTABLE_READABLE:
    return "PN_SELECTABLE_READABLE";
  case PN_SELECTABLE_WRITABLE:
    return "PN_SELECTABLE_WRITABLE";
  case PN_SELECTABLE_ERROR:
    return "PN_SELECTABLE_ERROR";
  case PN_SELECTABLE_EXPIRED:
    return "PN_SELECTABLE_EXPIRED";
  case PN_SELECTABLE_FINAL:
    return "PN_SELECTABLE_FINAL";
   case PN_CONNECTION_WAKE:
    return "PN_CONNECTION_WAKE";
   case PN_LISTENER_CLOSE:
    return "PN_LISTENER_CLOSE";
   case PN_PROACTOR_INTERRUPT:
    return "PN_PROACTOR_INTERRUPT";
   case PN_PROACTOR_TIMEOUT:
    return "PN_PROACTOR_TIMEOUT";
   case PN_PROACTOR_INACTIVE:
    return "PN_PROACTOR_INACTIVE";
   default:
    return "PN_UNKNOWN";
  }
  return NULL;
}

pn_event_t *pn_event_batch_next(pn_event_batch_t *batch) {
  return batch->next_event(batch);
}
